ambry-replication/src/main/java/com/github/ambry/replication/BackupCheckerThread.java

```diff
@@ -0,0 +1,165 @@
+  }
+
+  /**
+   * This method checks if a temporary blob is present in the backup.
+   * The replica could be lagging behind the backup and the blob could be permanent or even deleted from the backup.
+   * The replica should catch up with its peers. Nonetheless, this check will catch such replicas as well.
+   * @param validMessageDetectionInputStream Stream of valid blob IDs
+   * @param remoteReplicaInfo The remote replica that is being replicated from
+   */
+  @Override
+  protected void applyPut(MessageSievingInputStream validMessageDetectionInputStream,
+      RemoteReplicaInfo remoteReplicaInfo) {
+    EnumSet<MessageInfoType> acceptableBackupBlobStates = EnumSet.of(MessageInfoType.PUT);
+    EnumSet<StoreErrorCodes> acceptableStoreErrorCodes = EnumSet.noneOf(StoreErrorCodes.class);
+    for (MessageInfo onPremBlob: validMessageDetectionInputStream.getValidMessageInfoList()) {
+      checkBackup(onPremBlob, remoteReplicaInfo, acceptableBackupBlobStates, acceptableStoreErrorCodes);
+    }
+  }
+
+  /**
+   * This method checks if the blob is in local store in an acceptable state or encounters an acceptable error code
+   * TODO: check for more than just blob-states, check for expiry date for example
+   * @param onPremBlob the {@link MessageInfo} that will be checked for in the backup
+   * @param remoteReplicaInfo The remote replica that is being replicated from
+   * @param acceptableBackupBlobStates Acceptable states in which the blob must be present in the backup
+   * @param acceptableStoreErrorCodes Acceptable error codes when retrieving the blob from backup
+   */
+  protected void checkBackup(MessageInfo onPremBlob, RemoteReplicaInfo remoteReplicaInfo,
+      EnumSet<MessageInfoType> acceptableBackupBlobStates, EnumSet<StoreErrorCodes> acceptableStoreErrorCodes) {
+    try {
+      StoreInfo storeInfo = remoteReplicaInfo.getLocalStore().get(Collections.singletonList(onPremBlob.getStoreKey()),
+          EnumSet.noneOf(StoreGetOptions.class));
+      MessageInfo backupBlob = storeInfo.getMessageReadSetInfo().get(0);
+      if (!acceptableBackupBlobStates.contains(getBlobState(backupBlob))) {
+        logger.error(String.format("%s ON_PREM_STATE = %s, BACKUP_STATE = %s RemoteReplica = %s",
+            onPremBlob.getStoreKey(), getBlobState(onPremBlob), getBlobState(backupBlob), remoteReplicaInfo));
+      }
+    } catch (StoreException e) {
+      if (!acceptableStoreErrorCodes.contains(e.getErrorCode())) {
+        logger.error(String.format("%s ON_PREM_STATE = %s, BACKUP_STATE = %s RemoteReplica = %s",
+            onPremBlob.getStoreKey(), getBlobState(onPremBlob), e.getErrorCode(), remoteReplicaInfo));
+      }
+    }
+  }
+
+  /**
+   * Returns an enum corresponding to blob state
+   * @param messageInfo Blob state as message info object
+   * @return Blob state as enum
+   */
+  protected MessageInfoType getBlobState(MessageInfo messageInfo) {
```

Comment:
Yes, I need to think more about this. I have a feeling that this is correct though.

Reply:
"
can we move the MessageInfoType inside the MessageInfoType, not make a global enum/
"
This is not clear to me. What's the issue with making it a global enum ? If someone wants to use it in the future, they can. Also there is a TODO in MessageInfo class that this should be done.

Reply:
Sorry. I mean move MessageInfoType to BackupCheckerThread since only BackupCheckerThread is using it.
The reason is every blob has multiple states. With single state, I don't see the case we will use it in other code yet.

Reply:
I propose keeping it open to others. Its not getting in the way of anything